unikernel.ml

```diff
@@ -490,39 +497,186 @@ struct
+                                  ^ ";Path=/;HttpOnly=true"
+                                in
+                                let header_list =
+                                  [
+                                    ("Set-Cookie", cookie_value);
+                                    ("location", "/dashboard");
+                                  ]
+                                in
+                                Lwt.return
+                                  (reply ~header_list
+                                     ~content_type:"application/json" res)
+                            | Error (`Msg _msg) ->
+                                let res =
+                                  "{\"status\": 400, \"success\": false, \
+                                   \"message\": \"Something went wrong. Wait a \
+                                   few seconds and try again.\"}"
+                                in
+                                Lwt.return
+                                  (reply ~content_type:"application/json" res)))
+                    ))
+            | _ ->
+                let res =
+                  "{\"status\": 400, \"success\": false, \"message\": \"Bad \
+                   request method\"}"
+                in
+                Lwt.return (reply ~content_type:"application/json" res))
+        | "/api/login" -> (
+            let request = Httpaf.Reqd.request reqd in
+            match request.meth with
+            | `POST -> (
+                decode_request_body reqd >>= fun data ->
+                let json =
+                  try Ok (Yojson.Basic.from_string data)
+                  with Yojson.Json_error s -> Error (`Msg s)
+                in
+                match json with
+                | Error (`Msg s) ->
+                    Logs.warn (fun m -> m "Failed to parse JSON: %s" s);
+                    let res =
+                      "{\"status\": 400, \"message\": \"Bad request body\"}"
+                    in
+                    Lwt.return (reply ~content_type:"application/json" res)
+                | Ok json -> (
+                    let validate_email_re =
+                      Re.Perl.re "[a-zA-Z0-9.$_!]+@[a-zA-Z0-9]+\\.[a-z]{2,3}"
+                      |> Re.compile
+                    in
+                    let validate_user_input email password =
+                      if email = "" || password = "" then
+                        Error "All fields must be filled."
+                      else if not (Re.execp validate_email_re email) then
+                        Error "Invalid email address."
+                      else if String.length password < 8 then
+                        Error "Password must be at least 8 characters long."
+                      else Ok "Validation passed."
+                    in
+                    let email =
+                      json
+                      |> Yojson.Basic.Util.member "email"
+                      |> Yojson.Basic.to_string
```

Comment:
it is not clear to me what the `to_string` does, we want to ensure that the "email" thingy is a `` `String `` -- so in the end I guess we need a json_utils module (since we've already similar code in user_model and albatross_json...